src/java.base/share/classes/java/util/DualPivotQuicksort.java

```diff
@@ -618,6 +636,126 @@ private static void pushDown(int[] a, int p, int value, int low, int high) {
         a[p] = value;
     }
 
+    /**
+     * Sorts the specified range of the array using radix sort.
+     *
+     * @param a the array to be sorted
+     * @param low the index of the first element, inclusive, to be sorted
+     * @param high the index of the last element, exclusive, to be sorted
+     * @return true if finally sorted, false otherwise
+     */
+    static boolean radixSort(Sorter sorter, int[] a, int low, int high) {
+        int[] b; int offset = low;
+
+        if (sorter == null || (b = (int[]) sorter.b) == null) {
+            b = (int[]) tryAllocate(a, high - low);
+
+            if (b == null) {
+                return false;
+            }
+        } else {
+            offset = sorter.offset;
+        }
+
+        int start = low - offset;
+        int last = high - offset;
+
+        int[] count1 = new int[256];
+        int[] count2 = new int[256];
+        int[] count3 = new int[256];
+        int[] count4 = new int[256];
+
+        for (int i = low; i < high; ++i) {
+            count1[ a[i]         & 0xFF]--;
```

Comment:
Not a reviewer, but having recently implemented a radixsort myself I'm wondering what is the logic or benefit of decrementing and counting backwards here?

One thing I did differently, and I'm not fully sure is an optimization, is remembering the last bucket for each of the 4 counts. Checking whether the data is already sorted by that digit can then be done by checking `count[last_bucket] == size`, which avoids the first loop in `passLevel`. Again, not sure whether it is actually faster, maybe the two separate simple loops like here are better.

Reply:
@jhorstmann In counting backwards we perform comparing with 0 in a loop, Comparing with 0 may be faster than comparing with other value. In general, backward or forward moving is your favor.

Keeping last_bucket and use check count[last_bucket] == size sounds good, but we need to run benchmarking and compare. I think we will not see big difference because the size of count array is too small, 256 only, whereas we scan whole array with size at least 6K. The corner case when we can see max effect of using last_bucket is array with all equal elements (4 count arrays will have only one non-zero element). but such array will be caught by tryMerge method.